### PR TITLE
Add async dangling indices support

### PR DESCRIPTION
**Pull request by ywelsch**
Adds support for writing out dangling indices in an asynchronous way. Also provides an option to avoid writing out dangling indices at all.

Relates #48701

A follow-up will look at how the cluster state on data-only nodes can be written out asynchronously.  The main reason it is still written out in this PR is that the cluster state contains the cluster UUID, which allows data nodes to snap into a cluster.

### CODE REVIEW

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/1 (unrelated test failure)

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 1]
```diff
@@ -1487,4 +1529,48 @@ private void setIdFieldDataEnabled(boolean value) {
...
+            try {
+                danglingIndicesThreadPoolExecutor.execute(new AbstractRunnable() {
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.info(() -> new ParameterizedMessage("failed to write dangling indices state for index {}", index), e);
+                    }
+
+                    @Override
+                    protected void doRun() {
+                        final boolean exists = danglingIndicesToWrite.remove(index);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java]: Doing this before calling `indexService.writeDanglingIndicesInfo()` means that `allPendingDanglingIndicesWritten()` may return `true` while there's still a write in progress.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 2]
```diff
@@ -289,12 +306,25 @@ protected void closeInternal() {
                 }
             }
         };
+
+        final String nodeName = Objects.requireNonNull(Node.NODE_NAME_SETTING.get(settings));
+        danglingIndicesThreadPoolExecutor = EsExecutors.newScaling(
+            nodeName + "/" + DANGLING_INDICES_UPDATE_THREAD_NAME,
+            1, 1,
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java]: If `gateway.write_dangling_indices_info` is `false` I think we should avoid creating a thread here.

File: server/src/main/java/org/elasticsearch/gateway/MetaStateService.java [hunk 3]
```diff
@@ -264,15 +264,28 @@ public void cleanupIndex(Index index, long currentGeneration) {
...
+
     /**
      * Removes manifest file, global metadata and all index metadata
      */
     public void deleteAll() throws IOException {
-        MANIFEST_FORMAT.cleanupOldFiles(Long.MAX_VALUE, nodeEnv.nodeDataPaths());
+        // To ensure that the metadata is never reimported by loadFullStateBWC in case where the deletions here fail mid-way through,
+        // we first write an empty manifest file so that the indices become unreferenced, then clean up the indices, and only then delete
+        // the manifest file.
+        MANIFEST_FORMAT.writeAndCleanup(Manifest.empty(), nodeEnv.nodeDataPaths());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/MetaStateService.java]: Nice catch. Maybe call `unreferenceAll()` here instead?

File: server/src/main/java/org/elasticsearch/index/IndexService.java [hunk 4]
```diff
@@ -324,6 +327,29 @@ public synchronized void close(final String reason, boolean delete) throws IOExc
...
 
+    // method is synchronized so that IndexService can't be closed while we're writing out dangling indices information
+    public synchronized void writeDanglingIndicesInfo() {
+        if (closed.get()) {
+            return;
+        }
+        try {
+            IndexMetaData.FORMAT.writeAndCleanup(getMetaData(), nodeEnv.indexPaths(index()));
+        } catch (WriteStateException e) {
+            logger.info(() -> new ParameterizedMessage("failed to write dangling indices state for index {}", index()), e);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/IndexService.java]: I think a failure here indicates a bad enough problem to warrant a `WARN`-level log.

```suggestion
            logger.warn(() -> new ParameterizedMessage("failed to write dangling indices state for index {}", index()), e);
```

File: server/src/main/java/org/elasticsearch/index/IndexService.java [hunk 5]
```diff
@@ -324,6 +327,29 @@ public synchronized void close(final String reason, boolean delete) throws IOExc
...
+
+    // method is synchronized so that IndexService can't be closed while we're deleting dangling indices information
+    public synchronized void deleteDanglingIndicesInfo() {
+        if (closed.get()) {
+            return;
+        }
+        try {
+            MetaDataStateFormat.deleteMetaState(nodeEnv.indexPaths(index()));
+        } catch (IOException e) {
+            logger.info(() -> new ParameterizedMessage("failed to delete dangling indices state for index {}", index()), e);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/IndexService.java]: I think a failure here indicates a bad enough problem to warrant a `WARN`-level log.

```suggestion
            logger.warn(() -> new ParameterizedMessage("failed to delete dangling indices state for index {}", index()), e);
```

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java [hunk 6]
```diff
@@ -109,7 +109,11 @@ public void start(Settings settings, TransportService transportService, ClusterS
                             .build());
                     lucenePersistedState = new LucenePersistedState(
                         persistenceWriter, currentTerm, clusterState);
-                    metaStateService.deleteAll(); // delete legacy files
+                    if (DiscoveryNode.isDataNode(settings)) {
+                        metaStateService.unreferenceAll(); // unreference legacy files (only keep them for dangling indices functionality)
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: :+1: noting that if dangling indices functionality is disabled then we delete the metadata as the indices are allocated, so if any indices are genuinely dangling at this point then we keep their metadata around.

**Review by DaveCTurner — COMMENTED** [review]
Looks good. There's one issue I see with `allPendingDanglingIndicesWritten()` and I left a few more requests and comments.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/indices/IndicesService.java]: good catch. In an earlier iteration, I had `allPendingDanglingIndicesWritten` checking the executor as well, but I somehow thought this was unneeded later. I've reinstated the check in 3a7b656f5ce

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java (hunk 2 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/indices/IndicesService.java]: 👍 done in 3a7b656f5ce

File: server/src/main/java/org/elasticsearch/gateway/MetaStateService.java (hunk 3 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/gateway/MetaStateService.java]: ok, done in 3a7b656f5ce

File: server/src/main/java/org/elasticsearch/index/IndexService.java (hunk 4 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/IndexService.java]: 3a7b656f5ce

File: server/src/main/java/org/elasticsearch/index/IndexService.java (hunk 5 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/IndexService.java]: 3a7b656f5ce

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java (hunk 6 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: yup.

**Review by ywelsch — COMMENTED** [review]
Thanks for the review. I've addressed all comments. Can you have another look?

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**